### ISSUE TITLE
Bullets and upgrades

### ISSUE THREAD

**Issue by flip40** [DemoProductions/shmup#11]
This in part is to discuss how we will manage the different bullets and upgrades at a code level (to make it work as easily as possible). First though...

How do we want to have the game work? Will we have simple upgrades you can pick up that just upgrade a single bullet type (i.e. a single upgrade path)? Or should we have separate upgrades that can be picked up, like upgrades to bullets, upgrades to missiles (adding missiles on the first pickup), perhaps upgrades that change the bullet or missile type?

What do you guys think?

**Comment by stannardtyler** [comment]
Perhaps two types of power ups not including any life gain. Like one for constant that expires only upon death like a another cannon or backwards shooting, and one type that is set on a time limit expiration like a shield or huge/stronger bullets.

Other non enchanting power ups or drops could be health (if we don't do one hit kill) , extra lives, better armor? 

Maybe if we enjoy the game enough we can drop things that could further your ships strength with new weapons and ship parts. Jus an idea if we enjoy the game to even add that Rpg element lol.

**Comment by ghost** [comment]
I was thinking that maybe we can have different characters, and each character has a set type of weapon or bullet. If we implement the game that way, we would not have to worry about upgrades just for a certain weapon type. There might be different types of pickups that varies the amount of upgrade.

Or we can be like Cave Story. The player has a few weapons, but whatever weapon the player has equipped would be the one that gets the upgrade.

**Comment by flip40** [comment]
Yea, so the upgrade just gives a lvl 2 of their weapon and so on, or something like that. Or perhaps if you pick up a missile upgrade it will add missiles to the attacks (which again could vary between characters)... this sound on the right track?

**Comment by ghost** [comment]
Yeah something like that. Or we can have the upgrades be gradual so the player has to pick up something like energy orbs to level up the weapon gradually. Or larger orbs for more experience

**Comment by stannardtyler** [comment]
So added the bullet sprites for the player. They are blue and upgraded get a pixel bigger and change color. For now it's more so an upgrade of dmg rather than size or a spread like upgrade. 

![player-up bullet](https://cloud.githubusercontent.com/assets/15371385/17751524/5b290b02-647c-11e6-931f-59a47fe598e9.png)
![player-bullet](https://cloud.githubusercontent.com/assets/15371385/17751525/5b2be9c6-647c-11e6-96d3-5fb53dfff73b.png)

The player sprite sheet is more so the first bullet or two is more so when the bullet spawns, and the last two are the flashing animations.

The enemy ships just have the basic bullet sprite for now. The idea is the 3 bullets are a animation, constantly flashing to be seen with the dark space background we will prob lean towards.

![og bullet](https://cloud.githubusercontent.com/assets/15371385/17751530/60d30670-647c-11e6-8e0a-7c58fa4cabf2.png)
![new bullet](https://cloud.githubusercontent.com/assets/15371385/17751529/60d1b5ea-647c-11e6-81d0-71171e2308f2.png)

The only difference of the two, is one of the sprites rotates.

**Comment by stannardtyler** [comment]
I'll mock up some new bullet sprites, maybe a laser here and there. See how it looks in Unity and we can make a group choice on which bullets look or feel the best. For now I like the newer models that look more like a blast then a bullet. The green and blue that I posted above.

**Comment by flip40** [comment]
I agree, the 'blast' (green/blue) definitely looks nicer in my opinion. I'll see if I can add these bullet sprites in today, don't think I have done that yet and I got some free time (even more so, really, now that I've graduated, hurrah for no long wasted days in classes I don't need! \o/)

**Comment by flip40** [comment]
Oh, also make sure we get some enemy colored bullet sprites too! Hot colors for enemy bullets, cool colors for friendly bullets. Maybe we can shake that up in the long run, but for now that strikes me as the most practical in a UI sense (easy / natural to understand what will hurt you and what wont).

**Comment by stannardtyler** [comment]
I enjoy the concept of hot and cool colors to distinguish bullets, especially if we make a co op mode or something!
Also congratz man! I still got one more semester. Whats your next move after your schoolastic journey?
 

    On Tuesday, December 20, 2016 3:52 PM, Johnathon Ludeman <notifications@github.com> wrote:
 

 Oh, also make sure we get some enemy colored bullet sprites too! Hot colors for enemy bullets, cool colors for friendly bullets. Maybe we can shake that up in the long run, but for now that strikes me as the most practical in a UI sense (easy / natural to understand what will hurt you and what wont).—

**Comment by flip40** [comment]
Thanks! You are quite close yourself :D At the moment I start working full time at Kabam in January. Longer term is confusing. At some point within the next month or two I'll be working at a "new" company that is the spin-off of Kabam, as "Kabam" was sold off to another company (or rather, part of Kabam, though they inherit the name and we do not). From there it depends on if our new half-of-a-company gets sold or not, so given 6 months or so I have no idea where I'll be.

**Comment by stannardtyler** [comment]
Dude congratz! Yea I feel if I was doing animation/illustration program my portfolio would of got me in that Kabam program too, or at least have a better chance lol since my portfolio is mostly coding and websites I didn't have any badass illustrations! But i did learn alot and learned from a Blizzard presentation they had here on campus. Anyways you rock man and goodluck! 

What about you Adrian, you got still got a few semesters here or all done!?
 

    On Tuesday, December 20, 2016 4:10 PM, Johnathon Ludeman <notifications@github.com> wrote:
 

 Thanks! You are quite close yourself :D At the moment I start working full time at Kabam in January. Longer term is confusing. At some point within the next month or two I'll be working at a "new" company that is the spin-off of Kabam, as "Kabam" was sold off to another company (or rather, part of Kabam, though they inherit the name and we do not). From there it depends on if our new half-of-a-company gets sold or not, so given 6 months or so I have no idea where I'll be.—

**Comment by flip40** [comment]
Another thought I had on bullets, I think we should probably keep the average enemy bullet in a "round/square" shape, and not laser lines (with a bit of exception for specifically intending to shoot a laser beam across the entire screen).

My thought on this is in part the thought that most bullet-hell games seem to use round bullets, and it occurred to me that dodging circular points is far easier (or at least naturally understandable) than trying to dodge lines. The lines also make it oddly easier to run into the side of the bullet than be hit directly due to the length.

Of course, the bullets you have made already fit this perfectly (it is actually my default lasers sprites that are bad), just wanted to pass along thoughts on why we should probably keep bullets rounder / square shaped for the enemy. (and don't take the round / square _too_ literally, no need to be boring, just talking on overall dimension of the sprite and what it's expected hitbox would be).

Player bullets can be anything fun and interesting though, doesn't matter if the bots think our bullet sprites are hard to dodge.
